### ISSUE TITLE
Local cache for Open Street Map tiles

### ISSUE THREAD

**Issue by ispmarin** [SciTools/cartopy#732]
Natural Earth shapefiles are usually stored on the .local/share/cartopy folder. Is there a local folder where the tiles for OpenStreetMap img_tiles are stored?

**Comment by pelson** [comment]
Hi @ispmarin. Nice question!

Looking at https://github.com/SciTools/cartopy/blob/master/lib/cartopy/io/img_tiles.py#L184-L199 we can see that the image is fetched but never put to disk and is always an in-memory operation.

There isn't even an in-process cache for the tiles between rendering - making the whole thing pretty inefficient if you render tiles more than once (https://github.com/SciTools/cartopy/blob/master/lib/cartopy/mpl/geoaxes.py#L347-L350).

What's the usecase @ispmarin. How come you want the local cache?

**Comment by ispmarin** [comment]
Hey @pelson , thanks for your answer!

The use case is quite simple: I'm trying to use cartopy to plot several maps with a very bad internet connection. Usually the maps are using the same tiles and just changing some lines and points, so I suppose that it wouldn't be needed to keep downloading the same tiles over and over again. 

Other use case is that I would like to plot some maps using OSM tiles without internet at all, for several cases. I'm still struggling to use Natural Earth like this and so far there is no way to do it with OSM.

**Comment by ispmarin** [comment]
BTW, I tried to download more shapefiles from Natural Earth and put on the cache folder, renaming them, and cartopy is not capable of finding them. For example, a physical shapefile with 10m resolution would be named `10m_geographic_locations.shp`, with the other files, and I would try to plot using a feature like

    NaturalEarthFeature(category='physical', scale='10m', facecolor='none',
                             name='geographic_locations')

with no luck.

**Comment by pelson** [comment]
> BTW, I tried to download more shapefiles from Natural Earth and put on the cache folder, renaming them, and cartopy is not capable of finding them.

I'm going to treat those as separate issues (infact, if you're happy to, just open another issue on the tracker for that one).

> I'm trying to use cartopy to plot several maps with a very bad internet connection

Sounds reasonable. I know of plenty of situations in production environments where there is no connection *at all*. One option to get you going right now would be to subclass the tiler you want, and instead of fetching the tile from urllib into memory, it fetches it to disk, then loads from the cache. It should only represent <20 LOC. Want an example, or happy to have a go?

**Comment by ispmarin** [comment]
Hello again @pelson. I moved the Natural Earth to another issue, as you suggested. 

If you kindly could provide an example it would be great! 

Cheers

**Comment by pelson** [comment]
Something like:

```
import os
import types

import cartopy.io.img_tiles as img_tiles
import requests
import PIL


class CachedTiler(img_tiles.GoogleTiles):
    def __init__(self, tiler):
        self.tiler = tiler

    def __getattr__(self, name):
        # Mimic the tiler interface, but for methods, ensure that the "self"
        # that is passed through continues to be CachedTiler, and not the
        # contained tiler instance.
        attr = getattr(self.tiler, name)
        if isinstance(attr, types.MethodType):
            def wrapper(tiler, *args, **kwds):
                return attr(self, *args, **kwds)
            attr = wrapper
        return attr

    def get_image(self, tile):
        tileset_name = '{}'.format(self.tiler.__class__.__name__.lower())
        cache_dir = os.path.expanduser(os.path.join('~/', 'image_tiles', tileset_name))
        if not os.path.exists(cache_dir):
            os.makedirs(cache_dir)
        tile_fname = os.path.join(cache_dir, '_'.join(str(v) for v in tile) + '.png')
        if not os.path.exists(tile_fname):
            response = requests.get(self._image_url(tile),
                                    stream=True)

            with open(tile_fname, "wb") as fh:
                for chunk in response:
                    fh.write(chunk)
        with open(tile_fname, 'rb') as fh:
            img = PIL.Image.open(fh)
            img = img.convert(self.desired_tile_form)     
        return img, self.tileextent(tile), 'lower'
```

Which I'd use:

```
from cartopy.io.img_tiles import StamenTerrain
import matplotlib.pyplot as plt


actual_tiler = StamenTerrain()
tiler = CachedTiler(actual_tiler)

mercator = tiler.crs
ax = plt.axes(projection=mercator)
ax.set_extent([-90, -73, 22, 34])
ax.add_image(tiler, 6)
ax.coastlines('10m')
plt.show()
```

I remember that this is a dupe of #609 and that @ThomasLecocq would be interested in this solution too.

**Comment by ThomasLecocq** [comment]
Indeeeeeeeeeeeeeeeeeeeeeeed ! :)

**Comment by ispmarin** [comment]
Awesome! Thanks! 

I tested it offline and it worked. 

If I try to plot offline and the files are not cached I get the error

```
        382     """Return a single image, merging the given images."""
    383     if not tiles:
--> 384         raise ValueError('A non-empty list of tiles should '
    385                          'be provided to merge.')
    386     xset = [set(x) for i, x, y, _ in tiles]

ValueError: A non-empty list of tiles should be provided to merge.

<matplotlib.figure.Figure at 0x7fa8be2f5090>
```

That is not very informative but ok, I guess. Do you guys think that is is worth a pull request?

**Comment by pelson** [comment]
> Do you guys think that is is worth a pull request?

I can certainly see the utility of such a capability - so long as it was an opt-in thing, I'm ok with it being in cartopy. If you wanted to submit a PR @ispmarin, I'd happily merge once in a suitable state (it would need a unit test though). Some guidance on contributing: https://github.com/SciTools/cartopy/blob/master/CONTRIBUTING.md

:+1:

**Comment by ThomasLecocq** [comment]
:+1: top ! yes !

**Comment by darribas** [comment]
Sorry to jump in here, I think the issue is close enough that is probably easier this way. I'm trying to get a way of using `cartopy` I/O module and tiling functionality to arbitrarily pull down images (I haven't found any other easier way in Python) but not necessarily to plot them with `cartopy` but ideally read them (maybe on the fly) with `rasterio` for reprojection using the same CRS API as `geopandas`. I think the code given in this issue should be useful but I can't fully figure out how I could do that. Ideally, I'd want to do something like passing a `shapely` polygon (from `geopandas`, for example) and receiving back an image with the tile(s) that cover the extent of the polygon. Would this be very hard to do?

**Comment by pelson** [comment]
@darribas the [`image_for_domain`](https://github.com/SciTools/cartopy/blob/master/lib/cartopy/io/img_tiles.py#L76-L89) method on a tiler may be of interest to you. It takes a shapely geometry in Mercator coordinates and a desired z value, and returns a composite image (with metadata). Alternatively, there is also a `find_images` method which does a similar job, but doesn't do any of the image fetching for you - you could then do that manually with the `get_image` method and compose them as you wish.

The following code shows both of these options:

```
import cartopy.crs as ccrs
import cartopy.io.img_tiles as cimg_tiles
import shapely.geometry as sgeom


tiler = cimg_tiles.StamenTerrain()

lonlat_corners = (-10, 45), (10, 60)
z_target = 5

# Define the target in Mercator coordinates...
geo = ccrs.Geodetic()
x0, y0 = tiler.crs.transform_point(lonlat_corners[0][0], lonlat_corners[0][1], geo)
x1, y1 = tiler.crs.transform_point(lonlat_corners[1][0], lonlat_corners[1][1], geo)

target = sgeom.box(x0, y0, x1, y1)

tiles = list(tiler.find_images(target, target_z=z_target))

print('Using the following tiles: ', tiles)

import matplotlib.pyplot as plt

plt.figure()
plt.title('Single image')
single = tiler.get_image(tiles[0])
plt.imshow(single[0])

plt.figure()
plt.title('Composite image ({} tiles)'.format(len(tiles)))
composite = tiler.image_for_domain(target, target_z=z_target)
plt.imshow(composite[0], extent=composite[1], origin=composite[2])
plt.show()
```

**Comment by claudiodsf** [comment]
Hi, it looks like the solution provided by @pelson doesn't work anymore. Trying to use his code on carroty 0.14.2, I get the following error:

```
Traceback (most recent call last):
  File "test.py", line 51, in <module>
    ax.add_image(tiler, 6)
  File "/Users/claudio/anaconda3/lib/python3.6/site-packages/cartopy/mpl/geoaxes.py", line 283, in add_image
    assert not bool(args) and not bool(kwargs)
AssertionError
```

Delving into the `geoaxes.py` code, it looks like this is due to the fact that the method `image_for_domain` does not exist for `CachedTiler`.

Any suggestion?

**Comment by pelson** [comment]
@claudiodsf - could you double check? I'm able to get the example to work on v0.14 and v0.15.

**Comment by red5alex** [comment]
Getting the same AssertError as @claudiodsf  (with v0.15) now, but the code definitely worked recently. I had to reinstall Anaconda in the meantime, so maybe some updated libs may have changed it...

**Comment by stefanomattia** [comment]
Has this feature been included in v0.16? I've been sort of banned from google maps after running this code
```
figure, ax = plt.subplots(figsize=(15, 10))
request = cimgt.GoogleTiles(style="terrain")
ax = plt.axes(projection=request.crs)
ax.set_extent([94, 102, 0, 6])
ax.add_image(request, 8)
plt.show()
```
for a few times in my notebook. It would be neat to have the tiles cached as I'm going to fiddling a lot on the map.

**Comment by andy-watson** [comment]
I believe it maybe a Python 2/3 issue, `im_func` was replaced by `__func__`, altering reference in `__get_attr__()` method works for me with Python 3.

```
    def __getattr__(self, name):
         attr = getattr(self.tiler, name, None)
        if isinstance(attr, types.MethodType):
            attr = types.MethodType(attr.__func__, self)
        return attr
```

**Comment by claudiodsf** [comment]
I confirm that this is a Python 3 issue and that the solution provided by @andy-watson works for me.

Is there, thus, a way of having a Py2/3 compatible code?

**Comment by greglucas** [comment]
It looks like this was fixed in #1533
